### PR TITLE
[SPARK-24420][Build] Upgrade ASM to 6.1 to support JDK9+

### PR DESCRIPTION
**Pull request by dbtsai**
## What changes were proposed in this pull request?

Upgrade ASM to 6.1 to support JDK9+

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Review by felixcheung — APPROVED** [review]
cool
this will work with JDK 7 & 8?

**Comment by dbtsai** [comment]
@felixcheung Yes. All tests are passed with JDK8.

**Review by HyukjinKwon — APPROVED** [review]
Seems fine but just for clarification you plan to add a JDK9 specific tests and fix build settings for running it via Jenkins?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by dbtsai** [comment]
cc @srowen and @rxin for more eyes.

File: pom.xml [hunk 1]
```diff
@@ -313,13 +313,13 @@
         <artifactId>chill-java</artifactId>
         <version>${chill.version}</version>
       </dependency>
-      <!-- This artifact is a shaded version of ASM 5.0.4. The POM that was used to produce this
-           is at https://github.com/apache/geronimo-xbean/tree/xbean-4.4/xbean-asm5-shaded
+      <!-- This artifact is a shaded version of ASM 6.x. The POM that was used to produce this
```
Comment (srowen) [inline pom.xml]: Is this comment still true?

**Review by srowen — COMMENTED** [review]
It's a major version update in a dependency, so I think it's worth understanding what the key changes were? anything that looks like any potential issue? if tests all pass that's a good sign. And it's not something that *should* be used by apps, so I don't know if we must be concerned about the change to a leaked classpath. OK for Spark 3, probably for 2.4 too if there are no discernable problems.

File: pom.xml (hunk 1 above)
Comment (dbtsai) [reply pom.xml]: Yes, we still use a shaded version of asm 6.1.

**Comment by dbtsai** [comment]
@srowen From the release note,  http://asm.ow2.io/versions.html , the differences between 5.2 to 6.0 are

> Codebase migrated to gitlab (feature requests 317617, 317619, 317542)
Support for modules (JSR 376, feature request 317628)
Support for JDK specific module attributes (ModuleTarget, ModuleHashes and ModuleResolution).
ASM jars now contain debug info, asm-all jar removed (feature request 317622)
Support for removable features deleted (cf SIGNATURES, ANNOTATIONS, etc constants in ClassReader)
bug fixes
317793: AnalyzerAdapter calculates wrong type for AALOAD in case of null
317784: NegativeArraySizeException when resizing jump instruction in a class without frames
317780: IllegalArgumentException at CheckMethodAdapter.checkInternalName for non-Java-identifier name
317785: NullPointerException in MethodRemapper with 'chop' stack map frames and tree API
317786: VerifyError when added code pushes the offset of a conditional backward jump below -32768
317791: Invalid class files generated for large methods (regression introduced in 5.2)
317792: Fix the handling of inner classes with EXPAND_ASM_INSNS
Bug fix related to stack map frames

and the support of JSR 376 is what we need. Thanks.

**Comment by rxin** [comment]
What's driving this (is it java 9)? I'm in general scared by core library updates like this. Maybe Spark 3.0 is a good time (and we should just do it this year).

**Comment by dbtsai** [comment]
@rxin Mostly for Java 9+.

[ASM 6.x](https://mvnrepository.com/artifact/org.ow2.asm/asm/6.0/usages) has been proven in many projects such as FB Presto, Google Guice Core Library, CGLIB, Gradle, and Jetty Servlet Annotations, naming a few. 

Since it's a self-contained change and has couple bug fixes, how about we lay a good foundation in Spark 2.4 now? 

If it causes any instability or performance regression, we can easily revert it given we still have some time for 2.4 release.

**Review by srowen — APPROVED** [review]

**Comment by dbtsai** [comment]
There are three approvals from the committers, and the changes are pretty trivial to revert if we see any performance regression which is unlikely. To move thing forward, if there is no further objection, I'll merge it tomorrow. Thanks.

**Comment by rxin** [comment]
SGTM.

**Comment by dbtsai** [comment]
Thanks. Merged into master.

**Pull request closed without merging**